### PR TITLE
Use BackendCluster to represent OIDCProvider

### PR DESCRIPTION
**Pull request by zhaohuabing**
Using BackendCluster to represent OIDCProvider allows users to customize backend connection settings for OIDC providers, such as retry policy and tls setting.

Fix: https://github.com/envoyproxy/gateway/issues/3178

### CODE REVIEW

File: api/v1alpha1/oidc_types.go [hunk 1]
```diff
@@ -107,7 +107,8 @@ type OIDC struct {
 
 // OIDCProvider defines the OIDC Provider configuration.
 // +kubebuilder:validation:XValidation:rule="!has(self.backendRef)",message="BackendRefs must be used, backendRef is not supported."
-// +kubebuilder:validation:XValidation:rule="has(self.backendRefs)? self.backendRefs.size() > 1 : true",message="Only one backendRefs is allowed."
+// +kubebuilder:validation:XValidation:rule="has(self.backendSettings)? (has(self.backendSettings.retry)?(has(self.backendSettings.retry.perRetry)? !has(self.backendSettings.retry.perRetry.timeout):true):true):true",message="Retry timeout is not supported."
+// +kubebuilder:validation:XValidation:rule="has(self.backendSettings)? (has(self.backendSettings.retry)?(has(self.backendSettings.retry.retryOn)? !has(self.backendSettings.retry.retryOn.httpStatusCodes):true):true):true",message="HTTPStatusCodes is not supported."
```
Comment (zhaohuabing) [inline api/v1alpha1/oidc_types.go:111]: `Timeout` and `HTTPStatusCodes` RetryOn condition are not supported for non-route retry policy.

Non-route retry policy: https://www.envoyproxy.io/docs/envoy/latest/api-v3/config/core/v3/base.proto#envoy-v3-api-msg-config-core-v3-retrypolicy

Route retry policy: https://www.envoyproxy.io/docs/envoy/latest/api-v3/config/route/v3/route_components.proto#envoy-v3-api-msg-config-route-v3-retrypolicy

File: internal/gatewayapi/ext_service.go [hunk 2]
```diff
@@ -12,15 +12,66 @@ import (
...
 
 	egv1a1 "github.com/envoyproxy/gateway/api/v1alpha1"
 	"github.com/envoyproxy/gateway/internal/gatewayapi/resource"
 	"github.com/envoyproxy/gateway/internal/ir"
+	"github.com/envoyproxy/gateway/internal/utils"
 )
 
-// TODO: zhaohuabing combine this function with the one in the route translator
+// translateExtServiceBackendRefs translates external service backend references to route destinations.
+func (t *Translator) transalteExtServiceBackendRefs(
```
Comment (arkodg) [inline internal/gatewayapi/ext_service.go:26]:
```suggestion
func (t *Translator) translateExtServiceBackendRefs(
```

File: internal/gatewayapi/testdata/securitypolicy-with-oidc-custom-cookies.out.yaml [hunk 3]
```diff
@@ -188,6 +188,7 @@ xdsIR:
             name: securitypolicy/envoy-gateway/policy-for-gateway
             provider:
               authorizationEndpoint: https://accounts.google.com/o/oauth2/v2/auth
+              destination: null
```
Comment (arkodg) [inline internal/gatewayapi/testdata/securitypolicy-with-oidc-custom-cookies.out.yaml:191]: is this a bug ?

File: internal/gatewayapi/testdata/securitypolicy-with-oidc.in.yaml [hunk 4]
```diff
@@ -118,6 +120,8 @@ securityPolicies:
       name: httproute-1
     oidc:
       provider:
+        backendRefs:
```
Comment (arkodg) [inline internal/gatewayapi/testdata/securitypolicy-with-oidc.in.yaml:123]: this doesnt look valid

File: internal/xds/translator/testdata/in/xds-ir/oidc-backencluster-provider.yaml [hunk 5]
```diff
@@ -0,0 +1,60 @@
+http:
```
Comment (arkodg) [inline internal/xds/translator/testdata/in/xds-ir/oidc-backencluster-provider.yaml:1]: oidc-backend-cluster-provider.yaml

File: internal/gatewayapi/testdata/securitypolicy-with-oidc-custom-cookies.out.yaml (hunk 3 above)
Comment (zhaohuabing) [reply internal/gatewayapi/testdata/securitypolicy-with-oidc-custom-cookies.out.yaml:191]: destination should be set as `omitempty` as it's optional for OIDC provider.

File: internal/xds/translator/oidc.go [hunk 6]
```diff
@@ -210,9 +220,55 @@ func oauth2Config(oidc *ir.OIDC) (*oauth2v3.OAuth2, error) {
...
+		var rp *corev3.RetryPolicy
+		if rp, err = buildNonRouteRetryPolicy(oidc.Provider.Traffic.Retry); err != nil {
+			return nil, err
+		}
+		oauth2.Config.RetryPolicy = rp
+	}
 	return oauth2, nil
 }
 
+func buildNonRouteRetryPolicy(rr *ir.Retry) (*corev3.RetryPolicy, error) {
```
Comment (arkodg) [inline internal/xds/translator/oidc.go:234]: can't we reuse the existing helper func and just ignore/unset `PerRetry.Timeout` & `RetryOn.HTTPStatusCodes` ?

Comment (zhaohuabing) [reply internal/xds/translator/oidc.go:234]: They use different Retry types: `routev3.RetryPolicy` vs `corev3.RetryPolicy`.

**Review by arkodg — APPROVED** [review]
LGTM thanks !

**Comment by arkodg** [comment]
thanks @zhaohuabing !

**Review by zirain — APPROVED** [review]

**Pull request merged**